include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -835,7 +835,7 @@ __pattern_scan_copy(_ExecutionPolicy&& __exec, _Iterator1 __first, _Iterator1 __
         oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::write, _IteratorOrTuple>();
     auto __buf2 = __keep2(__output_first, __output_first + __n);
 
-    auto __res = __par_backend_hetero::__parallel_transform_scan(
+    auto __res = __par_backend_hetero::__parallel_transform_scan_multi_group(
```

Comment:
What used to be called `__parallel_transform_scan` did much more than just a generic transform scan: it received higher-order functions for a work-group level scan, a scan between work-groups and a global scan on the work-group scan's results. This by definition can not use a single-work group optimization because it by design is for multiple work groups. Most of the algorithms that called this function (such as `copy_if`) relied on overloading the higher-order functions and thus cannot easily be optimized for single work groups.

Because of this `__parallel_transform_scan` has been renamed to `__parallel_transform_scan_multi_group` to make it apparent that this is by design for multiple work groups. It also make it clearer that calling algorithms such as `copy_if` are explicitly relying on multiple work group behavior.

The `__parallel_transform_scan` algorithm now just does a

Reply:
I see. Thanks for clarifying. Should we comment this behavior in code then? I think this is quite different to the other algorithms. Especially since `copy_if` doesn't follow the sketch of the proposed design above.

Further, I think the naming of `__parallel_transform_scan`, `__parallel_transform_scan_single_group`, `__parallel_transform_scan_multi_group` could be improved to better distinguish them. Especially the naming of the latter two suggests that the only difference is how many work groups are used which doesn't seem to be the case.

Reply:
Quoting my answer to a previous question here, since it's relevant:

> After discussing with @rarutyun, we decided to rename `__parallel_transform_scan_multi_group` to `__parallel_transform_scan_base` to make it clearer that this function is a generic version that supports multiple algorithms.